include/unix/osd.h

```diff
@@ -45,6 +45,87 @@
 
 #define FI_DESTRUCTOR(func) static __attribute__((destructor)) void func
 
+#ifndef UNREFERENCED_PARAMETER
+#define OFI_UNUSED(var) (void)var
+#else
+#define OFI_UNUSED UNREFERENCED_PARAMETER
+#endif
+
+/* Atomic memory access macro */
+static inline
+long ofi_atomic_inc32(volatile long *ptr)
+{
+	return __sync_fetch_and_add(ptr, (long)1);
+}
+
+static inline
+long long ofi_atomic_inc64(volatile long long *ptr)
+{
+	return __sync_fetch_and_add(ptr, (long long)1);
+}
+
+static inline
+long ofi_atomic_dec32(volatile long *ptr)
+{
+	return __sync_fetch_and_sub(ptr, (long)1);
+}
+
+static inline
+long long ofi_atomic_dec64(volatile long long *ptr)
+{
+	return __sync_fetch_and_sub(ptr, (long long)1);
+}
+
+static inline
+long ofi_atomic_add32(volatile long *ptr, long value)
+{
+	return __sync_fetch_and_add(ptr, (long)value);
+}
+
+static inline
+long long ofi_atomic_add64(volatile long long *ptr, long long value)
+{
+	return __sync_fetch_and_add(ptr, (long)value);
+}
+
+static inline
+long ofi_atomic_sub32(volatile long *ptr, long value)
+{
+	return __sync_fetch_and_sub(ptr, (long)value);
+}
+
+static inline
+long long ofi_atomic_sub64(volatile long long *ptr, long long value)
+{
+	return __sync_fetch_and_sub(ptr, (long long)value);
+}
+
+static inline
+void* ofi_val_compare_and_swap_ptr(volatile void **ptr, void *ptr1, void *ptr2)
+{
+    return (void *)__sync_val_compare_and_swap(ptr, ptr1, ptr2);
+}
+
+static inline
+long ofi_val_compare_and_swap32(volatile long *ptr, long oldval, long newval)
+{
+	return __sync_val_compare_and_swap(ptr, oldval, newval);
+}
+
+static inline
+long long ofi_val_compare_and_swap64(volatile long long *ptr, long oldval, long newval)
+{
+	return __sync_val_compare_and_swap(ptr, oldval, newval);
+}
+
+static inline
+unsigned char ofi_val_compare_and_swap128(volatile long long *ptr, long long oldval,
+					  long long newval, long long *compare)
+{
+	OFI_UNUSED(compare);
+	return __sync_val_compare_and_swap((ptr), (oldval), (newval));
+}
```

Comment:
it was improper idea, but the idea is to use `__int128` for this. Seems, it was a bit incorrectly to rely on that since it may be a common code for Win OS and Unix-like OS